### ISSUE TITLE
Annotations on non-default constructors aren't found

### ISSUE THREAD

**Issue by mvoong** [google/j2objc#473]
The following class demonstrates that when isAnnotationPresent() is called on a non-default constructor,  it returns false. It correctly returns true on a default constructor.

```
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Constructor;

public class JavaTest {
    
    @Target({ElementType.CONSTRUCTOR})
    @Retention(RetentionPolicy.RUNTIME)
    public @interface Inject {
    }
    
    // This one works
    @Inject
    JavaTest() {
        
    }

//     This one doesn't work
//    @Inject
//    JavaTest(String hello) {
//        
//    }
    
    public void performTest() {
        Constructor<?>[] constructors = JavaTest.class.getDeclaredConstructors();
        assert constructors[0].isAnnotationPresent(Inject.class);
    }
}
```

I've done a bit of digging and it seems that getDeclaredAnnotations() in ExecutableMember.m calls -[JavaLangReflectConstructor internalName], which always returns "JavaTest". For the non-default constructor in the code sample, this should probably be returning "JavaTestWithNSString_".

**Comment by nbransby** [comment]
Any hints on this one if we were to attempt to fix ourselves?

**Comment by tomball** [comment]
When the second constructor is un-commented, this gets generated:

```objc
+ (IOSObjectArray *)__annotations_JavaTest {
  return ...
}

+ (IOSObjectArray *)__annotations_JavaTestWithNSString_ {
  return ...
}
```
Presently, constructor names start with the class name like Java, rather than "init" like Objective-C. As a quick hack, I'd therefore try changing Constructor.internalName to something like (warning, not verified):
```objc
- (NSString *)internalName {
  NSString *selectorName = NSStringFromSelector(selector_);
  return [selectorName stringByReplacingOccurrencesOfString:@"init"
                                                 withString:[class_ objcName]]];
}
```
For the real fix, I think we should change the generated annotation method name to just use a selector string (with underscores instead of colons), like is done for methods. We should also change Constructor's name methods to use metadata when available.

**Comment by mvoong** [comment]
Thanks Tom, the hack works great. We'll use this for now. Wish I could be familiar enough with the codebase to tackle your suggestions. @nbransby - sounds like fun?

**Comment by tomball** [comment]
Fixed in 0.9.6.
